### PR TITLE
Api: get/set rotationpolicy & rotate

### PR DESCRIPTION
**Pull request by qinl-li**
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i>

### Changelog
Add a changelog entry for this PR by answering the following questions:
  1. What's the purpose of the update?
      - [ ] new service onboarding
      - [ ] new API version
      - [ ] update existing version for new feature
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify
  2. When are you targeting to deploy the new service/feature to public regions? Please provide the date or, if the date is not yet available, the month.
  3. When do you expect to publish the swagger? Please provide date or, the the date is not yet available, the month.
  4. If updating an existing version, please select the specific langauge SDKs and CLIs that must be refreshed after the swagger is published.
      - [ ] SDK of .NET (need service team to ensure code readiness)
      - [ ] SDK of Python
      - [ ] SDK of Java
      - [ ] SDK of Js
      - [ ] SDK of Go
      - [ ] PowerShell
      - [ ] CLI
      - [ ] Terraform
      - [ ] No refresh required for updates in this PR

### Contribution checklist:
- [ ] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of "no breaking changes"
- [ ] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [ ] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist 

> **Applicability**: :warning: 
>
> If your changes encompass only the following scenarios, you should SKIP this section, as these scenarios do not require ARM review.
> - Change to data plane APIs 
> - Adding new properties 
> - All removals

Otherwise your PR may be subject to ARM review requirements. Complete the following:
- [ ] Check this box if any of the following apply to the PR so that label "WaitForARMFeedback" will be added automatically to begin ARM API Review. Failure to comply may result in delays to the manifest.
  - Adding a new service
  - Adding new API(s)
  - Adding a new API version
    -[ ] To review changes efficiently, ensure you are using OpenAPIHub to initialize the PR for adding a new version. More details, refer to the [wiki](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/208/OpenAPI-Hub-Adding-new-API-version).

- [ ] Ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them. 

### Breaking Change Review Checklist 
If any of the following scenarios apply to the PR, request approval from the Breaking Change Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 
- [ ] Removing API(s) in a stable version
- [ ] Removing properties in a stable version
- [ ] Removing API version(s) in a stable version
- [ ] Updating API in a stable or public preview version with Breaking Change Validation errors
- [ ] Updating API(s) in public preview over 1 year (refer to [Retirement of Previews](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37683/Retirement-of-Previews))

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Addition details on the process and office hours are on the [Breaking change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

### CODE REVIEW

File: specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/common.json [hunk 1]
```diff
@@ -69,6 +69,59 @@
...
+    "LifetimeActions": {
+      "properties": {
+        "trigger": {
+          "$ref": "#/definitions/LifetimeActionsTrigger",
+          "description": "The condition that will execute the action."
+        },
+        "action": {
+          "$ref": "#/definitions/LifetimeActionsType",
+          "description": "The action that will be executed."
+        }
```
Comment (heaths) [inline specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/common.json:82]: Which of these are required? I assume one or both, which should be added to a `required` array property in this model.

File: specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/common.json [hunk 2]
```diff
@@ -69,6 +69,59 @@
...
+    "LifetimeActionsTrigger": {
+      "properties": {
+        "timeAfterCreate": {
+          "type": "string",
+          "description": "Time after creation to attempt to rotate. It only applies to rotate. It will be in ISO 8601 duration format. Example: 90 days : \"P90D\" "
+        },
+        "timeBeforeExpiry": {
+          "type": "string",
+          "description": "Time before expiry to attempt to rotate or notify. It will be in ISO 8601 duration format. Example: 90 days : \"P90D\""
+        }
```
Comment (heaths) [inline specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/common.json:95]: Just as an FYI, we'll not currently be supporting ISO 8601 durations as anything other than a string in the Azure SDKs. We've discussed it at length and there are problems we can't easily resolve in most of our languages. We may do something in the future but not ready yet.

That's not to say don't use ISO 8601 durations - I prefer them to having to store a numerical value and unit type - but they'll just be strings in the SDK, so users are on the hook for formatting them. As such, you might consider adding a link to the IETF RFC for help. This text will, by default, be the doc comments for the SDK.

/cc @herveyw-msft as an FYI, though I think he was part of this conversation already.

File: specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/common.json [hunk 3]
```diff
@@ -69,6 +69,59 @@
...
+        "type": {
+          "type": "string",
+          "description": "The type of the action.",
+          "enum": [
+            "rotate",
+            "notify"
+          ],
+          "x-ms-enum": {
+            "name": "ActionType",
+            "modelAsString": false,
```
Comment (heaths) [inline specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/common.json:110]: If this will be in a response model (which it should), this should be `true`. Doesn't mean you have to support other models, but it will cause deserialization errors if the service returns more enums in the future. See https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md#use-extensible-enums. If you are 100% sure this will never change, this should be fine, though.

File: specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/secrets.json [hunk 4]
```diff
@@ -572,6 +572,137 @@
...
+        "operationId": "GetSecretsRotationPolicy",
+        "summary": "Lists the policy for a secret.",
+        "description": "The GetSecretRotationPolicy operation returns the specified secret policy resources in the specified secret vault. This operation requires the secret rotation policy get permission.",
+        "parameters": [
+          {
+            "name": "secret-name",
+            "in": "path",
+            "required": true,
+            "type": "string",
+            "description": "The name of the secret in a given key vault."
```
Comment (heaths) [inline specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/secrets.json:590]: Should capitalize product names, or maybe just say "vault" here.
```suggestion
            "description": "The name of the secret in a given Key Vault."
```

File: specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/secrets.json [hunk 5]
```diff
@@ -572,6 +572,137 @@
...
+        "operationId": "RotateSecret",
+        "summary": "Creates a new secret version, stores it, then returns secret value, attributes to the client.",
+        "description": "The operation will rotate the secret based on the secret rotation policy. It requires the secrets/rotate permission.",
+        "parameters": [
+          {
+            "name": "secret-name",
+            "in": "path",
+            "required": true,
+            "type": "string",
+            "pattern": "^[0-9a-zA-Z-]+$",
```
Comment (heaths) [inline specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/secrets.json:679]: Why is there a pattern here but nowhere else? Also, `secret-name` is repeated enough that maybe it should be in a `"parameters"` section and just `ref` it again and again. Just a thought, though.

File: specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/secrets.json [hunk 6]
```diff
@@ -877,6 +1008,70 @@
...
+          "description": "Provider settings in the form of key/value pairs."
+        },
+        "autoRotationSettings": {
+          "type": "object",
+          "additionalProperties": {
+            "type": "string"
+          },
+          "description": "Secret autorotation settings in the form of key/value pairs."
+        }
+      },
```
Comment (heaths) [inline specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/secrets.json:1052]: Before sure to document required properties. Same with any new model introduced here.

**Review by heaths — CHANGES_REQUESTED** [review]

File: specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/common.json (hunk 1 above)
Comment (jlichwa) [reply specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/common.json:82]: Just FYI. It is same as we had for keys, now we are adding it to secrets.

File: specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/common.json (hunk 2 above)
Comment (jlichwa) [reply specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/common.json:95]: We had discussion with review board and that was approved for keys and now we extend it to other objects in vault

Comment (heaths) [reply specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/common.json:95]: Right - I was part of those talks. This wasn't feedback so much as an FYI ("Just as an FYI..."). Since I thought we punted key rotation to 7.4-preview.1, I thought there might still be time to change this if you wanted to; though, you don't have to; like I said above, I prefer ISO 8601 durations myself - no reason to reinvent durations.

File: specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/common.json (hunk 1 above)
Comment (heaths) [reply specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/common.json:82]: Never mind. I misremembered we punted key rotation to 7.4 but see it is [destined for 7.3](https://github.com/Azure/azure-rest-api-specs/blob/main/specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.3-preview/keys.json). Note, however, that it's in servers' customers' best interest to be clear on which properties are required for future models.

File: specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/common.json (hunk 3 above)
Comment (heaths) [reply specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/common.json:110]: With drawn since this is the same as in 7.3, but please keep in mind for future models.

File: specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/secrets.json (hunk 5 above)
Comment (qinl-li) [reply specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/secrets.json]: The pattern exists in secret create api. I copied the parameter from there. I have moved it to a separated parameter and ref it in places where it was using

File: specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/secrets.json [hunk 7]
```diff
@@ -116,6 +116,94 @@
             "$ref": "./examples/DeleteSecret-example.json"
           }
         }
+      },
+      "get": {
```
Comment (heaths) [inline specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/secrets.json:111]: I think this is going to cause problems. The REST API already treats the `secret-version` as optional in the `/secrets/{secret-name}/{secret-version}` endpoint, and the SDK reflects this. If you add an operation like this, it will generate an overload that won't compile. Why this change? It seems unnecessary even from a pure REST client standpoint since `secret-version` is optional already.

File: specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/secrets.json [hunk 8]
```diff
@@ -573,95 +661,6 @@
         }
       }
     },
-    "/secrets/{secret-name}/rotationpolicy": {
```
Comment (heaths) [inline specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/secrets.json]: Why was this removed? Was it only so you could update the secret bundle with a policy? As noted above, that is going to create problems for the SDKs that have long since GA'd. I don't think all languages will be able to work around the issue.

I recommend keeping this, which would be consistent with Keys.

File: specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/secrets.json [hunk 9]
```diff
@@ -744,6 +743,21 @@
...
     },
+    "SecretBundleWithRotationPolicy": {
+      "allOf": [
+        {
+          "$ref": "#/definitions/SecretBundle"
+        }
+      ],
+      "properties": {        
+        "rotation_policy": {
+          "$ref": "#/definitions/SecretRotationPolicy",
```
Comment (heaths) [inline specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/secrets.json:704]: Why not just return a policy on SecretBundle like you do on KeyBundle? Not only can we not work around issues this might create for the SDK (see previous comments), but this is a departure from the pattern already established with Keys.

File: specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/secrets.json [hunk 10]
```diff
@@ -934,14 +948,19 @@
           "x-ms-client-name": "secretAttributes",
           "$ref": "#/definitions/SecretAttributes",
           "description": "The secret management attributes."
+        },
+        "rotation_policy": {
```
Comment (heaths) [inline specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/secrets.json:902]: Why is the policy on both the secret bundle and the secret attributes? Seems it should only be on one, just as it is with Keys.

**Review by heaths — CHANGES_REQUESTED** [review]
This seems like a departure from the established model pattern with Keys and will cause issues for some language SDKs. Can you use the same endpoints and model changes as with Keys?

/cc @jlichwa @herveyw-msft @jackrichins

**Comment by heaths** [comment]
@qinl-li overall, why does this implementation differ so much from what we already have for keys: https://github.com/Azure/azure-rest-api-specs/blob/main/specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.3-preview/keys.json. While some patterns may be unfortunate - like using `release_policy` instead of `releasePolicy` like you changed in this PR - we already established a pattern in a service that is about to GA. I think consistency within the service is more important.

Can we just keep the same endpoints (except keys/ -> secrets/) and model schema for consistency in the REST API? We can make the SDK consistent, but swagger review isn't just about the SDK but the REST APIs as well.

/cc @jlichwa @RandalliLama @herveyw-msft

**Comment by herveyw-msft** [comment]
+1 to @heaths comments.

File: specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/secrets.json (hunk 4 above)
Comment (qinl-li) [reply specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/secrets.json]: Updated with capitalized product names.

File: specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/secrets.json (hunk 7 above)
Comment (qinl-li) [reply specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/secrets.json]: oh. This explains why there is no get operation under path /secrets/{secret-name}. I have removed it now and updated it in the getting secret version operation

File: specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/secrets.json (hunk 8 above)
Comment (qinl-li) [reply specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/secrets.json]: Get/Set key rotation policy APIs were created to work with MHSM.  However, we don't have such dependencies in secret. 
Another reason that we want to combine the get/set rotation policy with the secret get/path operation is to make it work better with Azure Policy.  @jlichwa can add more details here :)

**Comment by jlichwa** [comment]
> @qinl-li overall, why does this implementation differ so much from what we already have for keys: https://github.com/Azure/azure-rest-api-specs/blob/main/specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.3-preview/keys.json. While some patterns may be unfortunate - like using `release_policy` instead of `releasePolicy` like you changed in this PR - we already established a pattern in a service that is about to GA. I think consistency within the service is more important.
> 
> Can we just keep the same endpoints (except keys/ -> secrets/) and model schema for consistency in the REST API? We can make the SDK consistent, but swagger review isn't just about the SDK but the REST APIs as well.
> 
> /cc @jlichwa @RandalliLama @herveyw-msft

Not sure about naming convention, but after detail review combining rotation policy simplifies REST API, permissions (similar to certificates), and number of calls. Also allows Azure Policy to enforce that secret cannot be created without rotation policy. We don't see scenario when we need separate API for rotation policy configuration management. 
Keys were special in this case for MHSM team requirement to keep rotation management separate, but we don't have that requirements for secrets.

File: specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/secrets.json (hunk 8 above)
Comment (heaths) [reply specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/secrets.json]: Are you saying MHSM does support this endpoint but KV doesn't? If that is the case, the endpoint needs to stay but perhaps KV can return an error - 501 not implemented - while MHSM supports it.

Comment (jackrichins) [reply specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/secrets.json]: AKV and MHSM support key rotation policy, but only AKV supports secrets.

/cc @dhruv-iyer

**Comment by JackTn** [comment]
Hi, @qinl-li. The PR has be closed for a long time and it's related branch still exist. Please tell me if you still need this branch or i will delete it in 14 days.

**Comment by JackTn** [comment]
Hi, @qinl-li. The PR has be closed for a long time and it's related branch still exist. Please tell me if you still need this branch or i will delete it in 14 days.

**Comment by JackTn** [comment]
Hi, @qinl-li. The PR has be closed for a long time and it's related branch still exist. Please tell me if you still need this branch or i will delete it in 14 days.
